File: Criacional/Singleton/Singleton.py
class ConexaoBanco:
    
    _instance = None 

    def __new__(cls, *args, **kwargs):
        """
        Método mágico chamado antes de __init__. 
        Controla a criação da instância.
        """
        # 1. Verifica se a instância única existe.
        if cls._instance is None:
            # 2. Se não existe, chama o método __new__ da classe base (object)
            # para criar e alocar o novo objeto.
            cls._instance = super(ConexaoBanco, cls).__new__(cls)
        
        # 3. Retorna a instância única (existente ou recém-criada).
        return cls._instance

    def __init__(self) -> None:
        """
        Inicializa a instância. O bloco __init__ só é executado na PRIMEIRA 
        criação (pois a instância é sempre retornada no __new__).
        """
        # Verifica se o objeto já foi inicializado (para evitar re-inicialização)
        if not hasattr(self, '_initialized'):
            self.settings = {
                "banco": "PostgreSQL", 
                "usuario": "Nety", 
                "senha": "Luzinete"
            }
            # Sinaliza que a inicialização foi concluída.
            self._initialized = True
            print("[LOG] Instância de Conexão Única inicializada.")
    
    def get_setting(self, key):
        return self.settings[key]

    def set_setting(self, key, value):
        self.settings[key] = value
        print(f"Alteração: configuração atualizada: {key}: {value}")

    def get_instance(self):
        if self._instance is None:
            self._instance = ConexaoBanco()
        return self._instance

print("Primeiro passo, instaciar a classe")

CONFIG = ConexaoBanco()

print(f"Dados da conexão: \nBanco: {CONFIG.get_setting('banco')}, \nUsuário: {CONFIG.get_setting('usuario')}, \nSenha: {CONFIG.get_setting('senha')}")

print("Segundo passo, alterar a instacia da classe")

CONFIG.set_setting('usuario', 'Silvia')

print(f"Dados da conexão: \nBanco: {CONFIG.get_setting('banco')}, \nUsuário: {CONFIG.get_setting('usuario')}, \nSenha: {CONFIG.get_setting('senha')}")

print("Terceiro passo, garantir que apenas uma instancia é criada")

CONFIG2 = ConexaoBanco()

CONFIG3 = ConexaoBanco().get_instance()

print(f"Instancia 1 é igual a instância 2? {CONFIG is CONFIG2}")
print(f"Instancia 2 é igual a instância 3? {CONFIG2 is CONFIG3}")
print(f"Instancia 1 é igual a instância 3? {CONFIG is CONFIG3}")

File: Criacional/Singleton/Singleton.js
// PRIMEIRA ETAPA - CLASSE A SER INSTANCIADA UNICAMENTE
class ConexaoBanco {
	// Instância estática da classe
	static instance = null;

	// Construtor bloqueado de instanciar a classe mais de uma vez
	constructor() {
		if (ConexaoBanco.instance) {
			return ConexaoBanco.instance;
		}

		// Caso não haja instância, cria uma nova instância e armazena-a
		this.settings = {
			banco: "PostgreSQL",
			usuario: "Nety",
			senha: "Luzinete",
		};

        ConexaoBanco.instance = this;
	}

	// Método para ler chave da instância
	getSetting(key) {
		return this.settings[key];
	}

	// Método para atualizar um valor da instância
	setSetting(key, value) {
		this.settings[key] = value;
		console.log(`Alteração: configuração atualizada: ${key}: ${value}`);
	}

	// Método para retornar a instância única
	static getInstance() {
		if (!this.instance) {
			this.instance = new ConexaoBanco();
		}
		return this.instance;
	}
}

// TESTE DA IMPLEMENTAÇÃO

console.log(`Primeiro passo, instaciar a classe`)
const config = new ConexaoBanco();
console.log(`Dados da conexão: \nBanco: ${config.getSetting('banco')}, \nUsuário: ${config.getSetting('usuario')}, \nSenha: ${config.getSetting('senha')}`)

console.log(`Segundo passo, alterar a instacia da classe`)

config.setSetting('usuario', 'Silvia')
console.log(`Dados da conexão: \nBanco: ${config.getSetting('banco')}, \nUsuário: ${config.getSetting('usuario')}, \nSenha: ${config.getSetting('senha')}`)

console.log(`Terceiro passo, garantir que apenas uma instancia é criada`)

const config2 = new ConexaoBanco();

const config3 = ConexaoBanco.getInstance()

console.log(`Instancia 1 é igual a instância 2? ${config === config2}`)
console.log(`Instancia 2 é igual a instância 3? ${config2 === config3}`)
console.log(`Instancia 1 é igual a instância 3? ${config === config3}`)
